### PR TITLE
fix overlapping definitions of `Base.active_module` and `REPL.active_module`

### LINKED ISSUE #54888: `Base.active_module()` design is very badly broken on master
In addition to existing milestone-blocking issues with `Base.active_module` function being broken and thread-unsafe, it can also outright crash the session if it gets into a slightly bad state:

```
[ Info: Precompiling REPLExt [e5eb5ef1-03cf-53a7-ae1d-5a66b08e832b] 
ERROR: TypeError: 
Stacktrace:
  [1] active_module()
    @ REPL ~/julia/usr/share/julia/stdlib/v1.12/REPL/src/REPL.jl:630
  [2] #invokelatest#2
    @ ./essentials.jl:1045 [inlined]
  [3] invokelatest
    @ ./essentials.jl:1042 [inlined]
  [4] active_module
    @ ./show.jl:519 [inlined]
  [5] 
SYSTEM (REPL): showing an error caused an error
ERROR: TypeError: 
Stacktrace:
  [1] active_module()
    @ REPL ~/julia/usr/share/julia/stdlib/v1.12/REPL/src/REPL.jl:630
  [2] #invokelatest#2
    @ ./essentials.jl:1045 [inlined]
  [3] invokelatest
    @ ./essentials.jl:1042 [inlined]
  [4] active_module
    @ ./show.jl:519 [inlined]
  [5] 
SYSTEM (REPL): caught exception of type TypeError while trying to handle a nested exception; giving up

julia> Base.active_module()

SYSTEM: caught exception of type :TypeError while trying to print a failed Task notice; giving up
ERROR: TaskFailedException
Stacktrace:
  [1] fatal: error thrown and no exception handler available.
TypeError(func=:typeassert, context="", expected=REPL.AbstractREPL, got=REPL.LineEditREPL(t=REPL.Terminals.TTYTerminal(term_type="xterm-256color", in_stream=Base.TTY(handle=0x000000012ef05da0, status=3, buffer=Base.GenericIOBuffer{Memory{UInt8}}(data=Memory{UInt8}(65536, 0x120600048)[0x28, 0x5b, 0x41, 0x00, 0x00, 0x00, 0x00, 0x00, 0x88, 0x55, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xc8, 0x55, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x23, 0x56, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x28, 0x56, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x83, 0x56, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xd0, 0x56, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x48, 0x57, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xd8, 0x57, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x48, 0x58, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xb8, 0x58, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x13, 0x59, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x60, 0x59, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xd8, 0x59, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x78, 0x5a, 0x00, 0x00, 0x00, 0x00, 0x00
# this keeps printing until the terminal crashes
```

### PR DESCRIPTION
**Pull request by JeffBezanson**
I'm hoping this will fix #54888 but I haven't been able to reproduce it so maybe somebody else can confirm.

### CODE REVIEW

File: base/show.jl [hunk 1]
```diff
@@ -514,9 +514,10 @@ function _show_default(io::IO, @nospecialize(x))
 end
 
 function active_module()
-    REPL = REPL_MODULE_REF[]
-    REPL === Base && return Main
-    return invokelatest(REPL.active_module)::Module
+    if !@isdefined(active_repl)
+        return Main
+    end
+    return invokelatest(active_module, active_repl)::Module
```
Comment (vtjnash) [inline base/show.jl:520]: I think you mean to just return it here, not call it

Comment (vtjnash) [reply base/show.jl:520]: In particular, anything that tries to use invokelatest will never work for this

Comment (JeffBezanson) [reply base/show.jl:520]: Do you mean there needs to be a case to handle `in_pure_context`?

**Comment by JeffBezanson** [comment]
This is hopefully ok now.

- Removed the strange function `is_exported_from_stdlib` that did not really do what it sounds like. These cases should just use `isvisible` IMO.
- Switch back to defaulting to Main instead of active_module. It's not ideal but needed for compatibility.

**Comment by JeffBezanson** [comment]
@nanosoldier `runtests(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/55316#issuecomment-2290264564) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/63493b2_vs_c7309d0/report.html) is available.

File: stdlib/REPL/src/REPL.jl [hunk 2]
```diff
@@ -17,6 +17,8 @@ module REPL
 Base.Experimental.@optlevel 1
 Base.Experimental.@max_methods 1
 
+import Base: active_module
```
Comment (KristofferC) [inline stdlib/REPL/src/REPL.jl:20]:
```suggestion
using Base: active_module
```
?

**Comment by JeffBezanson** [comment]
OK there was one package test failure that was actually expecting a qualified name, which is preferred since it is less state-dependent, so I switched function printing back to something more compatible with `is_exported_from_stdlib`. This is a good thing since is more stable than using Main or active_module.

**Pull request merged**